### PR TITLE
fix: emit close on stdin stream end

### PR DESCRIPTION
**Pull request by liz3**
We where not emitting the `close` event when done reading stdin data. It took me way too long to find this but i kept searching in zig/cpp code of the stream implementations...

Fixes: https://github.com/oven-sh/bun/issues/6713

### What does this PR do?

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

- [x] I included a test for the new code, or an existing test covers it

### CODE REVIEW

**Comment by Jarred-Sumner** [comment]
does this align with what node does? does node emit `close` before or after `end`? are they emitted on the same tick? or is there a `nextTick` between them? is it possible to emit `close` between `end` (or vice versa) and is it possible for node to end up emitting the event twice (do we need to handle that?)

**Comment by liz3** [comment]
@Jarred-Sumner 
image

**Comment by Jarred-Sumner** [comment]
i suggest also testing with inquirer

**Comment by liz3** [comment]
@Jarred-Sumner it was good you made me aware again because with this change `close` was being emited twice because of the internals `close` handler, meant making the logic more complex because we obviously don't want to emit `close` twice...
Node only sends `end` when the data is fully consumed, i tried getting `end` without close in node but didnt.
end/close are still emitted on the same tick but we only emit close once, through stream.destroy().

Il play around with inquirer.

**Comment by Jarred-Sumner** [comment]
is destroy correct there? I thought destroy can only be done once. What happens if you try to read from stdin twice this way?

**Comment by liz3** [comment]
> is destroy correct there? I thought destroy can only be done once. What happens if you try to read from stdin twice this way?

thats for what the boolean is though, il test what happens if you try to read again.
The destroy is responsible for actually emitting the `close` event.

**Comment by liz3** [comment]
@Jarred-Sumner it just resolves to null. `process.stdin.closed` and `process.stdin.destroyed` are true
image

File: src/js/builtins/ProcessObjectInternals.ts [hunk 1]
```diff
@@ -145,7 +147,11 @@ export function getStdinStream(fd) {
         }
       } else {
         stream.emit("end");
-        stream.pause();
+        if (!stream_destroyed) {
+          stream_destroyed = true;
+          stream.destroy();
+          unref();
+        }
```
Comment (jerome-benoit) [inline src/js/builtins/ProcessObjectInternals.ts:154]: Why don't you emit 'close' to avoid to duplicate code and add boolean variable?

Comment (liz3) [reply src/js/builtins/ProcessObjectInternals.ts:154]: because of the close handler below that.

**Comment by Jarred-Sumner** [comment]
Thank you

File: src/js/builtins/ProcessObjectInternals.ts (hunk 1 above)
Comment (jerome-benoit) [reply src/js/builtins/ProcessObjectInternals.ts:154]: Why destroy() on a stream is emitting the 'close' event? 
Semantically, destroying is orthogonal to emitting an event. The destroy() caller is responsible of emitting event accordingly given the context.

**Pull request merged**